### PR TITLE
SQL: make date format functions more strict

### PR DESCRIPTION
**Pull request by luigidellaquila**
Switching to `CLDR` locale provider (see https://github.com/elastic/elasticsearch/pull/110222) exposed some problems related to date conversion functions.
In particular, these functions are defined to mimic MySQL and PostgreSQL behavior, and tests that compare results with the corresponding queries in MySQL and PostgreSQL show a slightly different behavior with CLDR, eg. 
- days of week are returned as `Mon` instead of `Monday` with the ROOT locale; 
- calculation of week of the year considers slightly different defaults (eg. minimum number of days in the year for the first week).

This change makes the calculation more strict by 
- using ENGLISH locale 
- defining some conversion patterns not with the (apparently not so) equivalent Java pattern, but with the exact definitions given by the original specs (see https://dev.mysql.com/doc/refman/8.4/en/date-and-time-functions.html#function_date-format)

The problem is not present with COMPAT locale provider, the intention of this change is to fix failures introduced by https://github.com/elastic/elasticsearch/pull/110222, but I'm doing this as a separate PR to make sure that the new implementation is also good with COMPAT and we are not introducing regressions (ie. that it works fine with both `COMPAT` and `CLDR`)

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java [hunk 1]
```diff
@@ -59,19 +59,28 @@ class DateFormatter {
             new Builder().pattern("%s").javaPattern("ss").build(),
             new Builder().pattern("%T").javaPattern("HH:mm:ss").build(),
             new Builder().pattern("%U")
-                .javaFormat(t -> String.format(Locale.ROOT, "%02d", t.get(WeekFields.of(DayOfWeek.SUNDAY, 7).weekOfYear())))
+                .javaFormat(t -> String.format(Locale.ENGLISH, "%02d", t.get(WeekFields.of(DayOfWeek.SUNDAY, 7).weekOfYear())))
                 .build(),
-            new Builder().pattern("%u").javaFormat(t -> String.format(Locale.ROOT, "%02d", t.get(WeekFields.ISO.weekOfYear()))).build(),
+            new Builder().pattern("%u")
+                .javaFormat(t -> String.format(Locale.ENGLISH, "%02d", t.get(WeekFields.of(DayOfWeek.MONDAY, 4).weekOfYear())))
```
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java:65]: This is exactly how [MySQL spec](https://dev.mysql.com/doc/refman/8.4/en/date-and-time-functions.html#function_date-format) defines `%u`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java [hunk 2]
```diff
@@ -59,19 +59,28 @@ class DateFormatter {
...
+                .build(),
+
             new Builder().pattern("%V")
-                .javaFormat(t -> String.format(Locale.ROOT, "%02d", t.get(WeekFields.of(DayOfWeek.SUNDAY, 7).weekOfWeekBasedYear())))
+                .javaFormat(t -> String.format(Locale.ENGLISH, "%02d", t.get(WeekFields.of(DayOfWeek.SUNDAY, 7).weekOfWeekBasedYear())))
+                .build(),
+            new Builder().pattern("%v")
+                .javaFormat(t -> String.format(Locale.ENGLISH, "%02d", t.get(WeekFields.of(DayOfWeek.MONDAY, 4).weekOfWeekBasedYear())))
```
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java:72]: This is exactly how [MySQL spec](https://dev.mysql.com/doc/refman/8.4/en/date-and-time-functions.html#function_date-format) defines `%v`

**Comment by luigidellaquila** [comment]
Wait for https://github.com/elastic/elasticsearch/pull/112196 before merging

**Review by bpintea — APPROVED** [review]
🙏

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by thecoop** [comment]
@luigidellaquila Is it ok to backport this to 8.15 and 7.17?

**Comment by luigidellaquila** [comment]
@thecoop For 8.15 I don't see any problems.
For 7.17 I think we'll have to do a partial backport (I'm pretty sure [DATE_FORMAT](https://github.com/elastic/elasticsearch/pull/88388) function was never backported to 7.x)

**Comment by luigidellaquila** [comment]
I can take care of it tomorrow

**Pull request merged**